### ISSUE TITLE
AWQ and Gemma3 compatibility problem

### ISSUE THREAD

**Issue by BBC-Esq** [OpenNMT/CTranslate2#1956]
<html><body>
<hr>
<h2>Title: Gemma3 + AWQ produces NaN/overflow due to float16 limitations with large RMSNorm weights</h2>
<h3>Description</h3>
<p>Gemma3 models fail to produce meaningful output when using AWQ quantization in CTranslate2. The model outputs only <code>&lt;pad&gt;</code> tokens with a score of exactly <code>-65504.0</code> (the negative maximum value of float16), indicating numerical overflow.</p>
<p><strong>This issue is specific to AWQ quantization</strong> - the same Gemma3 model works correctly with int8 quantization in CTranslate2.</p>
<h3>Root Cause</h3>
<p>Gemma3 models have unusually large RMSNorm weights compared to other architectures. For example, in <code>gemma-3-1b-it</code>:</p>

LayerNorm | Min Weight | Max Weight
-- | -- | --
input_layernorm | 2.27 | 27.88
post_attention_layernorm | -1.41 | 51.25
pre_feedforward_layernorm | -284.00 | 0.25
post_feedforward_layernorm | -1.00 | 66.50

<p>These extreme values (especially -284) cause overflow when computed in float16, which has a maximum representable value of ~65504.</p>
<p>The AWQ path in CTranslate2 enforces float16 for all operations:</p>
<pre><code class="language-cpp">// From ops/awq/gemm.cc
if (a.dtype() != DataType::FLOAT16 &amp;&amp; b.dtype() != DataType::INT32)
    throw std::invalid_argument("Awq gemm is only supported for float16 input and int32 weight");
</code></pre>
<p>While the RMSNorm CUDA kernel does compute internally in float32, the output is cast back to float16, and subsequent operations (particularly the FFN's <code>silu(gate) * up</code> multiplication) overflow.</p>
<h3>Evidence</h3>
<p><strong>1. AWQ model produces garbage output:</strong></p>
<pre><code>Output tokens: ['&lt;pad&gt;', '&lt;pad&gt;', '&lt;pad&gt;', '&lt;pad&gt;', ...]
Score: -65504.0
</code></pre>
<p><strong>2. The same AWQ model works in PyTorch with float32 patching:</strong></p>
<p>When we wrap the MLP to compute <code>silu(gate) * up</code> in float32, the model works correctly:</p>
<pre><code class="language-python">class Gemma3MLPWrapper(nn.Module):
    def __init__(self, original_mlp):
        super().__init__()
        self.gate_proj = original_mlp.gate_proj
        self.up_proj = original_mlp.up_proj
        self.down_proj = original_mlp.down_proj
        self.act_fn = original_mlp.act_fn
    
    def forward(self, x):
        gate_output = self.gate_proj(x)
        up_output = self.up_proj(x)
        activated = self.act_fn(gate_output)
        # CRITICAL: Compute in float32 to avoid overflow
        intermediate = (activated.float() * up_output.float()).to(x.dtype)
        return self.down_proj(intermediate)
</code></pre>
<p>With this patch applied:</p>
<pre><code>AWQ logits range: [-29.33, 13.83]
AWQ PAD token rank: #110,051 (correct - not predicted)
Top prediction: '█' with logit 13.83
</code></pre>
<p><strong>3. Int8 quantization works fine:</strong></p>
<p>The same Gemma3 model converted with int8 quantization produces correct output, suggesting the issue is specific to the AWQ float16 enforcement.</p>
<h3>Affected Files</h3>
<ol>
<li><strong><code>ops/awq/gemm.cc</code></strong> - Enforces float16 input type</li>
<li><strong><code>ops/awq/gemv.cc</code></strong> - Same float16 enforcement</li>
<li><strong>FFN/MLP computation</strong> (location unknown) - The <code>silu(gate) * up</code> multiplication needs float32 intermediate computation</li>
</ol>
<h3>Proposed Solution</h3>
<p>The fix requires computing critical operations in float32 before casting back to float16:</p>
<p><strong>Option A: Model-specific flag for Gemma3</strong></p>
<p>Add a flag in the Gemma3 spec to enable float32 intermediate computations for models with large layernorm weights.</p>
<p><strong>Option B: Automatic detection</strong></p>
<p>Detect when layernorm weights exceed a threshold (e.g., abs &gt; 100) and automatically use float32 for intermediate computations.</p>
<p><strong>Option C: Modify AWQ kernels</strong></p>
<p>Update the AWQ GEMM/GEMV kernels to support float32 accumulation and output, similar to how the RMSNorm kernel already computes internally in float32:</p>
<pre><code class="language-cpp">// Current RMSNorm kernel already uses float32 internally:
float sum_squares = 0;
for (cuda::index_t i = threadIdx.x; i &lt; depth; i += blockDim.x)
    sum_squares += float(input[i]) * float(input[i]);  // float32
</code></pre>
<p>The same pattern should be applied to the FFN multiplication.</p>
<h3>Environment</h3>
<ul>
<li>CTranslate2 version: 4.6.2</li>
<li>PyTorch version: 2.9.0+cu128</li>
<li>CUDA: 12.8</li>
<li>GPU: NVIDIA RTX 4090</li>
<li>Model: gemma-3-1b-it quantized to AWQ (4-bit, group_size=128, version=gemm)</li>
</ul>

<hr>
</body></html>

**Comment by jordimas** [comment]
Thanks for reporting

Gemma 3 uses 4 RMSNorm layers per transformer block with zero-centered weights using a (1 + weight) scaling factor , which is different from standard implementations. 

Looking at the complexity of implementation, risk of introducing regressions in other models and performance impact across all models options A sounds to me the best. We can have a flag like requires_float32_intermediates that does not make so specific to Gemma3.

Let me know if you want to give a try to the fix or I can also work on this next week.

**Comment by vince62s** [comment]
FYI: you have to keep two distinct RMS if you want to stick to reproducible results (HF/vLLM do it too).
You can have a look at Eole here: https://github.com/eole-nlp/eole/blob/main/eole/modules/rmsnorm.py
